### PR TITLE
RavenDB-20652 Corax - compound sorting 

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20652 

### Additional description

Applying optimizations from https://github.com/ravendb/ravendb/pull/16453 into `SortingMultiMatch` and simplifying the API.

### Type of change

- Optimization

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
 Maybe?

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+
+            Span<long> buffer = stackalloc long[4] {_batchResults[x], _batchResults[y], -1, -1};
+            var swap = buffer[0] > buffer[1];
+            if (swap)
+                (buffer[0], buffer[1]) = (buffer[1], buffer[0]);
+            
+            _lookup.GetFor(buffer[..2], buffer[2..], BitConverter.DoubleToInt64Bits(double.MinValue));
+            return buffer[swap ? 3 : 2].CompareTo(buffer[swap ? 2 : 3]);
+        }
+    }
```
Comment (maciejaszyk) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: @ayende do that long keys persist the order in case of double or do I've to convert them into double?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        public static void IndirectSort<TComparer1, TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, Span<int> indexes,
+            UnmanagedSpan* batchTerms, TComparer1 comparer1, TComparer2 comparer2, TComparer3 comparer3)
+            where TComparer1 : struct, IComparer<UnmanagedSpan>, IEntryComparer
+            where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+        {
+            
+            if (match._orderMetadata[0].Ascending)
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:103]: why is this needed?

Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:103]: We've to do so since SortBatch can change the inner comparer (but only the first one). The best example is TermComparer where it creates `CompactKetComparer` inside.

I left it as similar to `SortingMatch` as possible since we still have to implement `SortUsingIndex`.
Take a look at `SortingMatch`:
https://github.com/ravendb/ravendb/blob/575ca9f1c432a23d14c20d6bdce13f89155a561f/src/Corax/Queries/SortingMatches/SortingMatch.Comparers.cs#L291-L320

**Comment by arekpalinski** [comment]
Please check failing tests

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:103]: @ayende is it fine then?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 3]
```diff
@@ -0,0 +1,701 @@
...
+
+        public void Init(ref SortingMultiMatch<TInner> match, UnmanagedSpan<long> batchResults, int comparerId)
+        {
+            _innerCmp.Init(ref match, batchResults, comparerId);
+        }
+
+        public void SortBatch<TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, LowLevelTransaction llt, PageLocator pageLocator, UnmanagedSpan<long> batchResults, Span<long> batchTermIds,
+            UnmanagedSpan* batchTerms, OrderMetadata[] orderMetadata, TComparer2 comparer2, TComparer3 comparer3) where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+        {
+            throw new NotImplementedException();
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: Shouldn't you forward here?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 4]
```diff
@@ -0,0 +1,701 @@
...
+        void Init(ref SortingMultiMatch<TInner> match, UnmanagedSpan<long> batchResults, int comparerId);
+
+        void SortBatch<TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, LowLevelTransaction llt, PageLocator pageLocator,
+            UnmanagedSpan<long> batchResults, Span<long> batchTermIds, UnmanagedSpan* batchTerms, OrderMetadata[] orderMetadata, TComparer2 comparer2,
+            TComparer3 comparer3)
+            where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer;
+    }
+
+    private struct Descending : IEntryComparer
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: What is this for?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 5]
```diff
@@ -0,0 +1,701 @@
...
+            return y.Double.CompareTo(x.Double);
+        }
+
+        public int Compare(int x, int y)
+        {
+            throw new NotSupportedException(ScoreComparerAsInnerExceptionMessage);
+        }
+    }
+
+    private struct CompactKeyComparer : IComparer<UnmanagedSpan>
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:160]: This is duplicate without any reason from SortingMatch, right?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 1 above)
Comment (ayende) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: You have to compare them using doubles, not long
The bit patterns are the same, the meaning is not

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+
+        private static long CopyTermPrefix(UnmanagedSpan item)
+        {
+            long l = 0;
+            Memory.Copy(&l, item.Address + 1 /* skip metadata byte */, Math.Min(6, item.Length - 1));
+            l = BinaryPrimitives.ReverseEndianness(l) >>> 1;
+            return l;
+        }
+
+        private static Span<int> SortByTerms<TComparer>(ref SortingMultiMatch<TInner> match, Span<long> buffer, UnmanagedSpan* batchTerms, bool isDescending,
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:254]: Need to create an issue to see if we can deal with this without so much code duplication

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    private static delegate*<ref SortingMultiMatch<TInner>, Span<long>, int> SortBy<TComparer1, TComparer2>(OrderMetadata[] orderMetadata)
+        where TComparer1 : struct, IEntryComparer, IComparer<UnmanagedSpan>
+        where TComparer2 : struct, IEntryComparer, IComparer<int>, IComparer<UnmanagedSpan>
+    {
+        if (orderMetadata.Length == 2)
+            return &Fill<TComparer1, TComparer2, NullComparer>;
+
+        return (orderMetadata[2].FieldType, orderMetadata[2].Ascending) switch
```
Comment (ayende) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:61]: How are you handling sorting by more? I'm not seeing the code for that?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 2 above)
Comment (ayende) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:103]: I'm pretty sure that we can't change the direction of the order, so why not check directly?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 4 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: Removed. I've made it to box interface but after consideration, I can just create boxed Dscending<T>.

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 3 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs]: Removed the  whole struct (now I'm using `Descedning<TCmp>`)

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:254]: Agreed, it's tricky since in `SortingMultiMatch` we need a little bit different method calls but maybe it's possible to make some parts common like `CompactKeyComparer`

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:103]: Well, this is an architectural issue from copying `SortingMatch` since it had to support `SortBatch` and `SortUsingIndex`.

For the first method, the generic type is only for creating the correct `IEntryComparer` to call `SortBatch`, Inside this we create 'new' instance of comparer (e.g. we completely change the comparer (the type) for `CompactKey` in TermComparer). 
This is why we had to wrap it in `IndirectComparer`.

However, in the case of `SortUsingIndex` we use that type immediately and use `Descending<TCmp>`. I thought we will work on `SortUsingIndex` in compound sorting soon so I have not changed the design to avoid additional work that will be removed anyway.

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+                cmp = comparerId switch
+                {
+                    0 => _cmp1.Compare(_terms[x], _terms[y]),
+                    1 => _cmp2.Compare(x, y),
+                    2 => _cmp3.Compare(x, y),
+                    _ => _nextComparers[comparerId - 3].Compare(x, y)
+                };
+            }
+
+            return cmp;
```
Comment (maciejaszyk) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:684]: Here is the code for sorting

**Comment by arekpalinski** [comment]
Tests are failing

File: src/Corax/Queries/SortingMatches/SortingMatch.BasicsComparers.cs [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static bool IsDigit(char c)
+    {
+        return c >= '0' && c <= '9';
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static byte ToUpper(byte c)
+    {
+        if (c >= 'a' && c <= 'z')
```
Comment (redknightlois) [inline src/Corax/Queries/SortingMatches/SortingMatch.BasicsComparers.cs:43]: For reference, for the analyzers I am building optimized vector aware version of many of the routines that are implemented here.

**Comment by redknightlois** [comment]
As a side note I like the way the code got simplified, can you also post the benchmark results after we know that it is done for the order-by?

**Comment by maciejaszyk** [comment]
> Tests are failing


Fixed related tests.
I believe failing tests are not related to the PR issue now.

File: src/Corax/Queries/SortingMatches/SortingMatch.BasicsComparers.cs (hunk 9 above)
Comment (arekpalinski) [reply src/Corax/Queries/SortingMatches/SortingMatch.BasicsComparers.cs:43]: So after both PRa are merged you we could use the optimized versions instead of those, right?

File: src/Corax/Queries/SortingMatches/IMatchComparer.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using Corax.Mappings;
+using Corax.Utils.Spatial;
+using Spatial4n.Shapes;
+
+namespace Corax.Queries.SortingMatches
+{
+    public enum MatchCompareFieldType : ushort
```
Comment (arekpalinski) [inline src/Corax/Queries/SortingMatches/IMatchComparer.cs]: Can we move it do a dedicated file?

File: src/Corax/Queries/SortingMatches/IMatchComparer.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    {        
+        MatchCompareFieldType FieldType { get; }
+
+        FieldMetadata Field { get; }
+
+        int CompareSequence(ReadOnlySpan<byte> sx, ReadOnlySpan<byte> sy);
+        int CompareNumerical<T>(T sx, T sy) where T : unmanaged, INumber<T>;
+    }
+
+    public interface ISpatialComparer : IMatchComparer
```
Comment (arekpalinski) [inline src/Corax/Queries/SortingMatches/IMatchComparer.cs]: Same here

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:160]: @maciejaszyk ping

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 6 above)
Comment (arekpalinski) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:254]: Did you create an issue for that?

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 7 above)
Comment (arekpalinski) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:61]: @maciejaszyk ping

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:254]: https://issues.hibernatingrhinos.com/issue/RavenDB-20771/Remove-code-duplication-in-SortingMatch-and-MultiSortingMatch

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 7 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:61]: https://github.com/ravendb/ravendb/pull/16679/files#r1238519949

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.cs [hunk 12]
```diff
@@ -2,715 +2,200 @@
...
+                    (true, MatchCompareFieldType.Integer) => new EntryComparerByLong(),
+                    (false, MatchCompareFieldType.Integer) => new Descending<EntryComparerByLong>(),
+                    
+                    (true, MatchCompareFieldType.Sequence) => new EntryComparerByTerm(),
+                    (false, MatchCompareFieldType.Sequence) => new Descending<EntryComparerByTerm>(),
+                    
+                    (true, MatchCompareFieldType.Spatial) => new EntryComparerBySpatial(),
+                    (false, MatchCompareFieldType.Spatial) => new Descending<EntryComparerBySpatial>(),
+                    _ => throw new NotSupportedException($"Ascending: {orderMetadata[metadataId].Ascending} | FieldType: {orderMetadata[metadataId].FieldType}.")
+                };
```
Comment (maciejaszyk) [inline src/Corax/Queries/SortingMatches/SortingMultiMatch.cs:73]: Here is the code when we've more than 3 comparers

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs (hunk 7 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs:61]: https://github.com/ravendb/ravendb/pull/16679/files#r1243278957

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 6 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:254]: https://issues.hibernatingrhinos.com/issue/RavenDB-20771/Remove-code-duplication-in-SortingMatch-and-MultiSortingMatch

File: src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs (hunk 5 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/SortingMultiMatch.Comparers.cs:160]: Yes, 
I created a ticket to take a look into all code duplications.
https://github.com/ravendb/ravendb/pull/16679#discussion_r1238223175

File: src/Corax/Queries/SortingMatches/IMatchComparer.cs (hunk 10 above)
Comment (maciejaszyk) [reply src/Corax/Queries/SortingMatches/IMatchComparer.cs]: done
